File: camerax_android/android/src/main/kotlin/dev/zeekr/camerax_android/core/resolutionselector/ResolutionSelectorImpl.kt
package dev.zeekr.camerax_android.core.resolutionselector

import androidx.camera.core.resolutionselector.AspectRatioStrategy
import androidx.camera.core.resolutionselector.ResolutionFilter
import androidx.camera.core.resolutionselector.ResolutionSelector
import androidx.camera.core.resolutionselector.ResolutionStrategy
import dev.zeekr.camerax_android.CameraXApiPigeonProxyApiRegistrar
import dev.zeekr.camerax_android.PigeonApiResolutionSelectorBuilderProxyApi
import dev.zeekr.camerax_android.PigeonApiResolutionSelectorProxyApi
import dev.zeekr.camerax_android.ResolutionSelectorModeApi

class ResolutionSelectorImpl(registrar: CameraXApiPigeonProxyApiRegistrar) :
    PigeonApiResolutionSelectorProxyApi(registrar) {
    class BuilderImpl(registrar: CameraXApiPigeonProxyApiRegistrar) :
        PigeonApiResolutionSelectorBuilderProxyApi(registrar) {
        override fun pigeon_defaultConstructor(): ResolutionSelector.Builder {
            return ResolutionSelector.Builder()
        }

        override fun setAllowedResolutionMode(
            pigeon_instance: ResolutionSelector.Builder, mode: ResolutionSelectorModeApi
        ): ResolutionSelector.Builder {
            return pigeon_instance.setAllowedResolutionMode(mode.impl)
        }

        override fun setAspectRatioStrategy(
            pigeon_instance: ResolutionSelector.Builder, aspectRatioStrategy: AspectRatioStrategy
        ): ResolutionSelector.Builder {
            return pigeon_instance.setAspectRatioStrategy(aspectRatioStrategy)
        }

        override fun setResolutionFilter(
            pigeon_instance: ResolutionSelector.Builder, resolutionFilter: ResolutionFilter
        ): ResolutionSelector.Builder {
            return pigeon_instance.setResolutionFilter(resolutionFilter)
        }

        override fun setResolutionStrategy(
            pigeon_instance: ResolutionSelector.Builder, resolutionStrategy: ResolutionStrategy
        ): ResolutionSelector.Builder {
            return pigeon_instance.setResolutionStrategy(resolutionStrategy)
        }

        override fun build(pigeon_instance: ResolutionSelector.Builder): ResolutionSelector {
            return pigeon_instance.build()
        }
    }

    override fun getAllowedResolutionMode(pigeon_instance: ResolutionSelector): ResolutionSelectorModeApi {
        return pigeon_instance.allowedResolutionMode.resolutionSelectorModeApi
    }

    override fun getAspectRatioStrategy(pigeon_instance: ResolutionSelector): AspectRatioStrategy {
        return pigeon_instance.aspectRatioStrategy
    }

    override fun getResolutionFilter(pigeon_instance: ResolutionSelector): ResolutionFilter? {
        return pigeon_instance.resolutionFilter
    }

    override fun getResolutionStrategy(pigeon_instance: ResolutionSelector): ResolutionStrategy? {
        return pigeon_instance.resolutionStrategy
    }
}

val Int.resolutionSelectorModeApi: ResolutionSelectorModeApi
    get() = when (this) {
        ResolutionSelector.PREFER_CAPTURE_RATE_OVER_HIGHER_RESOLUTION -> ResolutionSelectorModeApi.PREFER_CAPTURE_RATE_OVER_HIGHER_RESOLUTION
        ResolutionSelector.PREFER_HIGHER_RESOLUTION_OVER_CAPTURE_RATE -> ResolutionSelectorModeApi.PREFER_HIGHER_RESOLUTION_OVER_CAPTURE_RATE
        else -> throw NotImplementedError("Not implemented value: $this")
    }

val ResolutionSelectorModeApi.impl: Int
    get() = when (this) {
        ResolutionSelectorModeApi.PREFER_CAPTURE_RATE_OVER_HIGHER_RESOLUTION -> ResolutionSelector.PREFER_CAPTURE_RATE_OVER_HIGHER_RESOLUTION
        ResolutionSelectorModeApi.PREFER_HIGHER_RESOLUTION_OVER_CAPTURE_RATE -> ResolutionSelector.PREFER_HIGHER_RESOLUTION_OVER_CAPTURE_RATE
    }

File: camerax_ios/lib/src/core/resolution_selector/resolution_selector_impl.dart
import 'package:camerax_ios/src/api.dart';
import 'package:camerax_ios/src/core.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

final class ResolutionSelector$BuilderImpl
    implements ResolutionSelector$Builder {
  final ResolutionSelectorBuilderProxyApi api;

  ResolutionSelector$BuilderImpl.internal(this.api);

  @override
  Future<ResolutionSelector$Builder> setAllowedResolutionMode(
    ResolutionSelector$Mode mode,
  ) => throw UnimplementedError();

  @override
  Future<ResolutionSelector$Builder> setAspectRatioStrategy(
    AspectRatioStrategy aspectRatioStrategy,
  ) => throw UnimplementedError();

  @override
  Future<ResolutionSelector$Builder> setResolutionFilter(
    ResolutionFilter resolutionFilter,
  ) => throw UnimplementedError();

  @override
  Future<ResolutionSelector$Builder> setResolutionStrategy(
    ResolutionStrategy resolutionStrategy,
  ) => api.setResolutionStrategy(resolutionStrategy.api).then((e) => e.impl);

  @override
  Future<ResolutionSelector> build() => api.build().then((e) => e.impl);
}

final class ResolutionSelectorImpl implements ResolutionSelector {
  final ResolutionSelectorProxyApi api;

  ResolutionSelectorImpl.internal(this.api);

  @override
  Future<ResolutionSelector$Mode> getAllowedResolutionMode() =>
      throw UnimplementedError();

  @override
  Future<AspectRatioStrategy> getAspectRatioStrategy() =>
      throw UnimplementedError();

  @override
  Future<ResolutionFilter?> getResolutionFilter() => throw UnimplementedError();

  @override
  Future<ResolutionStrategy?> getResolutionStrategy() =>
      api.getResolutionStrategy().then((e) => e?.impl);
}

final class ResolutionSelectorChannelImpl extends ResolutionSelectorChannel {
  @override
  ResolutionSelector$Builder createBuilder() {
    final api = ResolutionSelectorBuilderProxyApi();
    return ResolutionSelector$BuilderImpl.internal(api);
  }
}

extension ResolutionSelector$ModeX on ResolutionSelector$Mode {
  ResolutionSelectorModeApi get api => ResolutionSelectorModeApi.values[index];
}

extension ResolutionSelectorModeApiX on ResolutionSelectorModeApi {
  ResolutionSelector$Mode get impl => ResolutionSelector$Mode.values[index];
}

extension ResolutionSelectorBuilderProxyApiX
    on ResolutionSelectorBuilderProxyApi {
  ResolutionSelector$Builder get impl =>
      ResolutionSelector$BuilderImpl.internal(this);
}

extension ResolutionSelectorX on ResolutionSelector {
  ResolutionSelectorProxyApi get api {
    final impl = this;
    if (impl is! ResolutionSelectorImpl) throw TypeError();
    return impl.api;
  }
}

extension ResolutionSelectorProxyApiX on ResolutionSelectorProxyApi {
  ResolutionSelector get impl => ResolutionSelectorImpl.internal(this);
}
